### DISCUSSION TITLE
Documentation: creating custom events

### DISCUSSION THREAD

**Discussion by Satyam** [lit/lit#2583] [Ideas]
Since events are the standard way to propagate information up the rendered tree, may I suggest in the documentation section about [dispatching CustomEvents](https://lit.dev/docs/components/events/#standard-custom-events) that an improved way to create and use custom subclasses of `CustomEvent` be documented?   

It took me a while for me to find ways to do it in proper TypeScript fashion, and being such a critical part of upstream signalling in an app, it might be good to document it. The most popular solution seems to be to type cast the listener to be an `EventListener` (appending `as EventListener`) but that leaves you clueless as to the contents of `ev.detail`.

```js
export const LOGIN_EVENT: 'loginEvent' = 'loginEvent' as const;

type LoginEventDetail = User | null;

export class LoginEvent extends CustomEvent<LoginEventDetail> {
  constructor(detail: LoginEventDetail) {
    super(LOGIN_EVENT, { detail });
  }
}

declare global {
  interface WindowEventMap {
    [LOGIN_EVENT]: LoginEvent;
  }
}
```
Where `User` is the type or interface for the object containing the current logged in user info.

To use it you do, once the user has been authenticated and the user info saved in `currentUser`:

```js
window.dispatchEvent(new LoginEvent(currentUser));
```

And, at the other end, the event listener can be declared like:
```js
loginEventHandler= (ev: LoginEvent) => {
    this._currentUserName = ev.detail.userName;
}
```

By doing it in this manner, VsCode IntelliSense offers suggestion upon typing `ev.detail.`.

Perhaps some part of the declaration of the custom `CustomEvent` could be contained in a decorator?

Thanks

**Comment by zandaqo** [comment]
I'd just go with `Event`, it's much simpler:
```typescript
export class LoginEvent extends Event {
  constructor(public user: User) {
    super('loginEvent');
  }
}
declare global {
  interface HTMLElementEventMap {
    loginEvent: LoginEvent;
  }
}
```
There is nothing special about `CustomEvent` anymore, it was used to pass data in olden days when we couldn't extend DOM classes and didn't even have classes in JS, that's not the case now.

Edit: As for documenting the practice of extending `*EventMap` interfaces in TypeScript, IMO that's more of a TS problem, but I agree it won't hurt Lit to have at least in the given TS example.

**Reply by justinfagnani** [reply]
I agree here. If you're going to actually define a subclass, you might as well subclass `Event` rather than `CustomEvent`. `CustomEvent` seems most useful if you don't want to boilerplate of a subclass.

**Comment by Satyam** [comment]
Let me expand on this issue:

There are two possible classes you can extend to create your events.  `Event` and `CustomEvent`.  Formally, `CustomEvent` is the *proper* one to use because since you are only supposed to store your custom data in the `detail` property, and that one is documented and reserved for this use, no future changes to the base `Event` interface would affect it.  It also eliminates any risk of collision with any of the existing properties by having one of your property names match that of an existing property.

That said, there is nothing wrong with adding properties to the base `Event` or `CustomEvent`, **as long as you don't use any of the existing property names**.   All the `*target` properties are clearly to be avoided as they can be changed when the event propagates, specially by [event retargeting](https://lit.dev/docs/components/events/#shadowdom-retargeting).

Thus, you can either do (as shown above):

```js
export class LoginEvent extends CustomEvent<LoginEventDetail> {
  constructor(detail: LoginEventDetail) {
    super(LOGIN_EVENT, { detail });
  }
}
```
or:
```js
export class LoginEvent extends Event {
  currentUser: LoginEventDetail;
  constructor(currentUser: LoginEventDetail) {
    super(LOGIN_EVENT, , { composed: true, bubbles: true });
    this.currentUser = currentUser;
  }
}
```

At the other end you have two possible ways of declaring the event depending on where you want to listen to it, namely, globally on `window` or at the `HTMLElement` level or even both.

For example, following the `LoginEvent` example, you would want it to be broadcast widely to all components in a page, even those that are not on the same branch of the component that dispatched the event.   Thus, add it to `WindowEventMap`:

```js
declare global {
  interface WindowEventMap {
    [LOGIN_EVENT]: LoginEvent;
  }
}
```
And you want to fire it on the `window` object:

```js
window.dispatchEvent(LOGIN_EVENT, new LoginEvent(currentUser));
```

and add the listener like this:

```js
connectedCallback() {
  super.connectedCallback();
  window.addEventListener(LOGIN_EVENT, this._handleLogin);
}
disconnectedCallback() {
  window.removeEventListener(LOGIN_EVENT, this._handleLogin);
  super.disconnectedCallback();
}
```
However, some events don't merit propagating all the way to the top, you might want to contain it to a limited sub-tree.  For example, in a component such as an accordion, clicking on any of the accordion panels is of interest to the overall accordion container or to the page that contains it, but you might want to capture halfway and stop its propagation to the top, thus, you might  want to be able to listen to it at any node in the tree.  

Also, if you listen to it at the `window` level, then you would be getting accordion panel events from panels from other accordions elsewhere in the page and your listener would have to check the target of the event to ensure the panel does belong to your accordion container.  

Thus you might want to declare it like this (you use `HTMLElementEventMap` instead of `WindowEventMap`):

```js
declare global {
  interface HTMLElementEventMap {
    [ACCORDION_EVENT]: AccordionEvent;
  }
}
```
and fire it like

```js
this.dispatchEvent(ACCORDION_EVENT, new AccordionEvent(panelName, panelOpen));
```

Then, you can set a listener to it in the constructor of any containing component:

```js
constructor() {
    super();
    this.addEventListener(ACCORDION_EVENT, this._handleAccordionPanelEvent);
  }
```
You don't need to remove that listener.

You can also add the listener like this:

```js
  protected override render() {
    return html`<div @accordion-event=${this._handleAccordionPanelEvent}></div>`;
  }
```

So, I think this covers all the options, and I believe they would be a good addition to the documentation.
Thanks

**Reply by zandaqo** [reply]
> Formally, CustomEvent is the proper one to use because since you are only supposed to store your custom data in the detail property, and that one is documented and reserved for this use, no future changes to the base Event interface would affect it.

While true, it's hard to imagine Event interface changing in the future for it to be a concern. Also, we are adding extra properties to DOM derived classes all the time in web components: custom elements are extensions of HTMLElement and in an unlikely event of HTMLElement getting new properties, those might conflict and be overshadowed by our own properties.

**Reply by Satyam** [reply]
Anyway, as the heading for this issue says, this is a **Documentation** issue and it is mostly the addition of the newly created event types to either of the `XxxEventMap` tables that I care about.  

In the same vein, I don't mean to discuss whether using `Event` or `CustomEvent` is the way to go, my intention is that both are documented and developers make their choices.   

Actually, I use `Event` and not `CustomEvent` myself, but I made that choice and others should be able to do likewise with all the information at hand.

**Reply by justinfagnani** [reply]
> Formally, CustomEvent is the proper one to use because since you are only supposed to store your custom data in the detail property

This is not true. `CustomEvent` exists because you didn't used to be able to subclass `Event`. Now that you can, there's not much reason to use `CustomEvent`. Many built-in classes are subclassable, including HTMLElement for our custom elements, Array, Map, etc. There is the standard subclassing hazzard, but CustomEvent was not created to avoid it.

**Reply by Satyam** [reply]
Whatever the historical reason for the existence of `CustomEvent`, which I can't argue since that is simply what I've heard, there is a further documentation issue.     There are far more examples using `CustomEvent` than `Event`.   There are none of subclassing either and neither are any examples on how to add it to any of the Event Maps.  So, whichever the current preferred method of firing events might be, it would be great to be have those issues covered.   Really, there is no point arguing here whether `Event` or `CustomEvent` is the way to go if the whole documentation points the other way (occurrences of `CustomEvent` out-votes `Event`), and misses those crucial bits, namely,  subclassing (whichever base class) and adding it to the corresponding EventMap.   Notice that the title refers to "custom events" not to `CustomEvent`.
